src/Renci.SshNet/Sftp/SftpFileStream.cs

```diff
@@ -267,6 +292,110 @@ internal SftpFileStream(ISftpSession session, string path, FileMode mode, FileAc
+                case FileAccess.ReadWrite:
+                    flags |= Flags.Read;
+                    flags |= Flags.Write;
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException("access");
+            }
+
+            if ((access & FileAccess.Read) != 0 && mode == FileMode.Append)
+            {
+                throw new ArgumentException(string.Format("{0} mode can be requested only when combined with write-only access.", mode.ToString("G")));
+            }
+
+            if ((access & FileAccess.Write) == 0)
+            {
+                if (mode == FileMode.Create || mode == FileMode.CreateNew || mode == FileMode.Truncate || mode == FileMode.Append)
+                {
+                    throw new ArgumentException(string.Format("Combining {0}: {1} with {2}: {3} is invalid.",
+                        typeof(FileMode).Name,
+                        mode,
+                        typeof(FileAccess).Name,
+                        access));
+                }
+            }
+
+            byte[] handle = null;
+
+            switch (mode)
+            {
+                case FileMode.Append:
+                    flags |= Flags.Append | Flags.CreateNewOrOpen;
+                    break;
+                case FileMode.Create:
+                    try
+                    {
+                        handle = await session.OpenAsync(path, flags | Flags.Truncate, cancellationToken).ConfigureAwait(false);
+                        flags |= Flags.Truncate;
+                    }
+                    catch (TaskCanceledException)// ToDo: when (cancellationToken.IsCancellationRequested)
+                    {
+                        throw;
+                    }
+                    catch (Exception)
+                    {
+                        flags |= Flags.CreateNew;
+                    }
+                    break;
+                case FileMode.CreateNew:
+                    flags |= Flags.CreateNew;
+                    break;
+                case FileMode.Open:
+                    break;
+                case FileMode.OpenOrCreate:
+                    flags |= Flags.CreateNewOrOpen;
+                    break;
+                case FileMode.Truncate:
+                    flags |= Flags.Truncate;
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException("mode");
+            }
+
+            if (handle == null)
+                handle = await session.OpenAsync(path, flags, cancellationToken).ConfigureAwait(false);
+
+            long position = 0;
+            if (mode == FileMode.Append)
+            {
+                try
+                {
+                    var attributes = await session.RequestFStatAsync(handle, cancellationToken).ConfigureAwait(false);
+                    position = attributes.Size;
+                }
+                catch
+                {
+                    session.RequestClose(handle);
```

Comment:
Use async variant.

Reply:
Well, this one is between a rock and a hard place: Because this is inside a catch block I cannot await an async operation. Then maybe I could just start the async operation and not await the task, but this could lead to race conditions when the caller wants to dispose the session and the ```RequestCloseAsync``` hasn't completed yet.